stdlib/builtins.pyi

```diff
@@ -940,23 +939,33 @@ class bool(int):
     def __invert__(self) -> int: ...
 
 @final
-class slice(Generic[_StartT, _StopT, _StepT]):
+class slice(Generic[_StartT_co, _StopT_co, _StepT_co]):
     @property
-    def start(self) -> _StartT: ...
+    def start(self) -> _StartT_co: ...
     @property
-    def step(self) -> _StepT: ...
+    def step(self) -> _StepT_co: ...
     @property
-    def stop(self) -> _StopT: ...
+    def stop(self) -> _StopT_co: ...
+    # generic slice --------------------------------------------------------------------
     @overload
-    def __new__(cls, stop: int | None, /) -> slice[int | MaybeNone, int | MaybeNone, int | MaybeNone]: ...
+    def __new__(cls, start: None, stop: None = None, step: None = None, /) -> slice[Any, Any, Any]: ...
```

Comment:
How bad do you think it would be to infer `slice[int | Any, int | Any, int | Any]` for this case (which is what we have on `main` -- take a look at the first overload)? A slice that has only `int`s or `None`s is by far the most common case; I think it's reasonable to infer the most common case for a slice if you haven't given it any explicit type annotations

```suggestion
    def __new__(cls, start: int | None, stop: int | None = None, step: int | None = None, /) -> slice[int | Any, int | Any, int | Any]: ...
```

Reply:
Though I suppose ideally you probably still want `slice(1, 1)` to result in `slice[int, int, int | Any]` rather than `slice[int | Any, int | Any, int | Any]`.

Reply:
I'm sure this is a good idea. 

First, I don't really understand what the presumed advantage of this is?

Secondly, an example I have in mind when thinking about this is slicing time series data. Under this convention, wouldn't something like

```python
selection: slice[datetime | None] = slice(None)
```

produce a type error?

Reply:
pyright doesn't like it:

Code sample in [pyright playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcAUCQMYA2AhgM62EDaAKgPrlgC6AXCVP1AAmAU2BRWrFMIDuEgBRVaASigBaAHxQARAUZE4UAD6YUMTlu5QAdDbICoAAQTgEwvKXsixtGGARzaYUpgFQ0oNg5LGysyAGJwxGFtXX0jEzMtTAYUMHw6WiQ0FGoAI0ok3yFhKmoQYUFYROTGHxA0gDkwKXMSAA9LXVaOruFOKABeQjkVKHjAGXISIA)

```python
from typing import Any

class A[T_co]:
    def __new__(cls) -> "A[Any | int]": ...

    @property
    def stop(self) -> T_co: ...

# Type "A[Any | int]" is not assignable to declared type "A[str | None]"
x: A[str | None] = A()  # ❌

```

Reply:
neither does mypy: https://gist.github.com/mypy-play/2812f15c3f3cac751bdb1771e63c44ba

Reply:
> pyright doesn't like it:

Fair enough. Well, we can leave it for now. I can always experiment in a followup PR. I'd still love some test cases for these uses, though, if they're things you want us to make sure we don't break in the future!